File: interpreter.py
import re as regex
import time

class ccpu:
    def __init__(self) -> None:
        self.regs: dict[str, int] = {
            "rax": 0,
            "rbx": 0,
            "rcx": 0,
            "rsp": 0,
            "rbp": 0,
            "r0": 0,
            "r1": 0,
            "r2": 0,
            "r3": 0,
            "r4": 0,
            "r5": 0,
            "r6": 0,
            "r7": 0,
            
        }
        # instantiate the ALU as part of the CPU
        self.alu = self.calu()
        self.mem = self.cmem()

    class cmem:
        def __init__(self) -> None:
            self.MEMORY_SIZE = 1000010  # 1 mb
            self.memory = bytearray(self.MEMORY_SIZE)

        def read_mem(self,adr:int,size:int) -> int:
            return int.from_bytes(self.memory[adr:adr+size], "little")
        
        def write_mem(self,val:int,adr:int,size:int) -> None:
            self.memory[adr:adr+size] = val.to_bytes(size, "little")

    class calu:
        def __init__(self) -> None:
            self.flags: dict[str, bool] = {
                "zf": False,  # zero flag
                "sf": False
            }
            self.branch = self.cbranch(self)

        def inc(self, a:str) -> None:
            write_val(a,get_val(a)+1)

        def dec(self, a:str) -> None:
            write_val(a,get_val(a)-1)

        def neg(self, a:str) -> None:
            write_val(a,get_val(a)*-1)

        def add(self, a:str, b:str) -> None:
            result: int = get_val(a) + get_val(b)
            self.set_flags(result)
            write_val(a,result)

        def sub(self, a:str, b:str) -> None:
            result: int = get_val(a) - get_val(b)
            self.set_flags(result)
            write_val(a,result)

        def mul(self, a:str, b:str) -> None:
            result: int = get_val(a) * get_val(b)
            self.set_flags(result)
            write_val(a,result)

        def div(self, a:str, b:str) -> None:
            result: int = get_val(a) // get_val(b)
            self.set_flags(result)
            write_val(a,result)

        def shl(self, a:str, b:str) -> None:
            write_val(a,get_val(a) << get_val(b))

        def shr(self, a:str, b:str) -> None:
            write_val(a,get_val(a) >> get_val(b))

        def andistrc(self,a:str, b:str) -> None:
            write_val(a,get_val(a) & get_val(b))

        def notistrc(self,a:str) -> None:
            write_val(a,~get_val(a))

        def oristrc(self,a:str, b:str) -> None:
            write_val(a,get_val(a) | get_val(b))

        def xoristrc(self,a:str, b:str) -> None:
            write_val(a,get_val(a) ^ get_val(b))
        
        def set_flags(self,a:int) -> None:
            self.flags["zf"] = (a == 0)
            self.flags["sf"] = (a < 0)  


        def test(self, a, b) -> None:
            self.set_flags(a&b)

        def cmp(self, a, b) -> None:
            self.set_flags(a-b)

        class cbranch:
            def __init__(self, alu) -> None:
                self.alu = alu   # keep reference to ALU

            def jmp(self, a) -> None:
                global i
                i = get_val(a)

            def jne(self, a) -> None:
                if not self.alu.flags["zf"]:   
                    global i
                    i = get_val(a)

            def je(self, a) -> None:
                if self.alu.flags["zf"]:   
                    global i
                    i = get_val(a)

            def jge(self, a) -> None:
                if not self.alu.flags["gf"]:   
                    global i
                    i = get_val(a)
                    





cpu=ccpu()


def eval_address(expr: str) -> int:
    """
    Evaluate x86-style address expressions:
    - [rax]
    - [rbp-8]
    - [rax+rbx*2+16]
    - [rbx*4+32]
    """
    expr = expr.replace(" ", "")

    # Replace registers with their numeric values
    for reg in cpu.regs:
        expr = regex.sub(rf"\b{reg}\b", str(cpu.regs[reg]), expr)

    # Validate: only numbers, + - * allowed now
    if not regex.fullmatch(r"[0-9+\-*()]+", expr):
        raise SyntaxError(f"Invalid memory expression: {expr}")

    # Safe eval
    return eval(expr)

sizes: dict[str, int] = {"byte":1,"word":2,"dword":4,"qword":8}

def get_val(tex:str) -> int:
    if tex.isnumeric():
        return int(tex)

    elif tex in cpu.regs.keys():
        return cpu.regs[tex]
    
    elif tex.endswith(":"):
        if tex in lables.keys():
            return lables[tex]
        else:
            raise SyntaxError(f"lable not defined: {tex}")
        
    elif tex.split(" ",2)[0] in sizes.keys():
        meml: list[str] = tex.split(" ")
        exp: str=meml[2][1:-1]
        size = sizes[meml[0]]
        expr = meml[2][1:-1]      # strip [ ]
        addr = eval_address(expr)
        return cpu.mem.read_mem(addr, size)
    else:
        raise SyntaxError("balls")

    

def write_val(were:str,val:int) -> None:
    if were in cpu.regs.keys():
        cpu.regs[were] = val
    elif were.split(" ",2)[0] in sizes.keys():
        meml: list[str] = were.split(" ",2)
        exp: str=meml[2][1:-1]
        cpu.mem.write_mem(val,get_val(exp),sizes[meml[0]])

    else:
        raise SyntaxError("ligma")




        
            
comands: list[str] =["mov","add","sub","mul","div","cmp","jmp","jne","jeq","lea","pri"] #pri is definitly asm yeah


lables :dict[str,int]= {}


with open("input.asm","r") as file:
    text: list[str]=file.readlines()

def parse(liner:str )->list[str]:
    lin: list[str] = regex.split(r'[,\s]+', line)
    nlin: list[str] = []
    i=0
    while i  < len(lin):
        if lin[i] in sizes.keys() and i < (len(lin)-2):
            sumline: str = lin[i]+" "
            sumline+=lin[i+1]+" "
            sumline+=lin[i+2]
            nlin.append(sumline)
            i+=3
        elif lin[i].startswith(";"):
            return nlin

        elif lin[i]:
            
            nlin.append(lin[i])
            i+=1

        else:
            i+=1

    return nlin

code=[]

for line in text:
    laa: list[str]=parse(line)
    if laa:
        code.append(laa)

print(code)

# Pass 1: collect labels
for idx, line in enumerate(code):
    if line[0].endswith(":"):
        if line[0] in lables:
            raise SyntaxError(f"Duplicate label: {line[0]}")
        lables[line[0]] = idx

# Pass 2: execution loop

start_time: float = time.perf_counter()

i:int=0
while i < len(code) :
    ccl:str=code[i][0]  #curent code line

    #print(f"isp: {i}")

    
    match ccl:
        case "mov":
            write_val(code[i][1],get_val(code[i][2]))

        case "lea":
            write_val(code[i][1],eval_address(code[i][2][1:-1]))


        case "inc":
            cpu.alu.inc(code[i][1])

        case "neg":
            cpu.alu.neg(code[i][1])

        case "shl":
            cpu.alu.shl(code[i][1],code[i][2])

        case "shr":
            cpu.alu.shr(code[i][1],code[i][2])

        case "add":
            cpu.alu.add(code[i][1],code[i][2])

        case "sub":
            cpu.alu.sub(code[i][1],code[i][2])

        case "mul":
            cpu.alu.mul(code[i][1],code[i][2])

        case "div":
            cpu.alu.div(code[i][1],code[i][2])

        case "not":
            cpu.alu.notistrc(code[i][1])

        case "and":
            cpu.alu.andistrc(code[i][1],code[i][2])

        case "or":
            cpu.alu.andistrc(code[i][1],code[i][2])

        

        


        case "cmp":
            cpu.alu.cmp(get_val(code[i][1]),get_val(code[i][2]))

        case "jmp":
            cpu.alu.branch.jmp(code[i][1])

        case "jne":
            cpu.alu.branch.jne(code[i][1])

        case "je":
            cpu.alu.branch.je(code[i][1])

        case "jmp":
            cpu.alu.branch.jmp(code[i][1])


        case "pri":
            print(get_val(code[i][1]))
    

    i+=1

end_time: float = time.perf_counter()
print(f"Execution time: {end_time - start_time:.5f} seconds")

File: interpreter.c
#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#include <stdbool.h>
#include <stdlib.h>
#include <time.h>
#include "mem.h"



enum { IMM=0, REG=1, MEM=2 };

int64_t regs[]={0,0,0,0,0,0,0,0,0,0,0,0};

uint8_t mem[1000010];

bool zf= false;
bool sf= false;


/*
"zf": False,  # zero flag
"gf": False,  # greater flag
"lf": False,  # less flag
*/

int64_t read_mem(size_t adr, uint8_t size) {
    //printf("read: addr: %zu size:%u",adr,size);
    switch (size) {
        case 0: return *(int8_t*)(mem + adr);
        case 1: return *(int16_t*)(mem + adr);
        case 2: return *(int32_t*)(mem + adr);
        case 3: return *(int64_t*)(mem + adr);
        default: return 0; // error
    }
}

void write_mem(size_t adr, uint8_t size, int64_t val) {
    //printf("write: addr: %zu size:%u val: %lld\n",adr,size,val);
    switch (size) {
        case 0: *(int8_t*)(mem + adr) = (int8_t)val; break;
        case 1: *(int16_t*)(mem + adr) = (int16_t)val; break;
        case 2: *(int32_t*)(mem + adr) = (int32_t)val; break;
        case 3: *(int64_t*)(mem + adr) = (int64_t)val; break;
    }
}

static inline void set_flags(int result){
    zf= !result;
    sf= (result<0);
}


static inline int64_t get_val(uint8_t mode, int val) {
    if (mode == IMM) return val;
    if (mode == REG) return regs[val];
    if (mode == MEM){
        uint32_t u = (uint32_t)val;
        uint8_t size = (uint8_t)(u >> 30);
        uint32_t num = u & ((1u << 30) - 1);

        return read_mem(mem_table[num](),size);
    }
    return 1;
}
static inline void write_val(uint8_t mode, int val, int64_t wval) {
    if (mode == REG) regs[val] = wval;
    if (mode == MEM){
        uint32_t u = (uint32_t)val;
        uint8_t size = (uint8_t)(u >> 30);
        uint32_t num = u & ((1u << 30) - 1);

        write_mem(mem_table[num](),size,wval);
    }
}

typedef struct {
    uint8_t op;   // opcode
    uint8_t m1;   // operand 1 mode
    uint8_t m2;   // operand 2 mode
    int32_t v1;   // operand 1 value
    int32_t v2;   // operand 2 value
} Instr;


Instr *decode(uint8_t *code, size_t size, size_t *out_count) {
    size_t n = size / 11; // since each instr = 1 + (1+4)*2 = 11 bytes
    Instr *prog = malloc(n * sizeof(Instr));
    size_t ip = 0;
    for (size_t i = 0; i < n; i++) {
        prog[i].op = code[ip++];
        prog[i].m1 = code[ip++];
        prog[i].v1 = *(int32_t*)(code + ip); ip += 4;
        prog[i].m2 = code[ip++];
        prog[i].v2 = *(int32_t*)(code + ip); ip += 4;
    }
    *out_count = n;
    return prog;
}




int main(){
  
    
    FILE *f = fopen("C:/Users/Levi/Documents/GitHub/asm_interpreter/program.bin", "rb");
    if (!f) return -1;
    fseek(f, 0, SEEK_END);
    long size = ftell(f);
    rewind(f);


    uint8_t *code = malloc(size);
    fread(code, 1, size, f);
    fclose(f);

    // now `code` holds the whole bytecode stream

    

    int result = 0;

    

    /*
    "rax": 0,
    "rbx": 1,
    "rcx": 2,
    "rsp": 3,
    "rbp": 4,
    "r0" : 5,
    "r1" : 6,
    "r2" : 7,
    "r3" : 8,
    "r4" : 9,
    "r5" : 10,
    "r6" : 11,
    "r7" : 12}
    */

    size_t n_instrs;
    Instr *prog = decode(code, size, &n_instrs);

    clock_t start = clock();
    size_t ip = 0;
    while (ip < n_instrs ){

        Instr in = prog[ip];

        //printf("op: %d, ip:%d \n",in.op,ip);

        

        switch (in.op){
            case 0:
                //mov
                

                write_val(in.m1,in.v1,get_val(in.m2,in.v2));

                break;
            case 1:
                //inc


                write_val(in.m1,in.v1,get_val(in.m1,in.v1)+1);
                
                break;
            
            case 2:
                //dec

                write_val(in.m1,in.v1,get_val(in.m1,in.v1)-1);
                
                break;

            case 3:
                //neg
                

                write_val(in.m1,in.v1,get_val(in.m1,in.v1)*-1);
                
                break;
            case 4:
                //shl
                

                result = get_val(in.m1,in.v1) << get_val(in.m2,in.v2);
                set_flags(result);

               
                write_val(in.m1,in.v1,result);
                
                break;
            case 5:
                //shr
                

                result = get_val(in.m1,in.v1) >> get_val(in.m2,in.v2);
                set_flags(result);

               
                write_val(in.m1,in.v1,result);

                
                break;
            case 6:
                //add
                

                result = get_val(in.m1,in.v1)+get_val(in.m2,in.v2);
                set_flags(result);

               
                write_val(in.m1,in.v1,result);

                
                break;
            case 7:
                //sub
                

                result = get_val(in.m1,in.v1)-get_val(in.m2,in.v2);
                set_flags(result);

               
                write_val(in.m1,in.v1,result);

                
                break;
            case 8:
                //mul
                

                result = get_val(in.m1,in.v1)*get_val(in.m2,in.v2);
                set_flags(result);

               
                write_val(in.m1,in.v1,result);

                
                break;
            case 9:
                //div
                

                result = get_val(in.m1,in.v1)/get_val(in.m2,in.v2);
                set_flags(result);

               
                write_val(in.m1,in.v1,result);

                
                break;

            case 10:
                //cmp


                set_flags(get_val(in.m1,in.v1)-get_val(in.m2,in.v2));
                
                break;

            case 11:
                //jmp

                ip = (size_t)get_val(in.m1,in.v1);
                continue;
                
                
                break;
            case 12:
                //jne
                if (!zf){
                    ip = (size_t)get_val(in.m1,in.v1);
                    continue;
                }
                
                
                break;
            
            case 13:
                //je
                if (zf){
                    ip = (size_t)get_val(in.m1,in.v1);
                    continue;
                }
                
                
                
                break;

            case 14:
                //jg
                if (!zf){
                    ip = (size_t)get_val(in.m1,in.v1);
                    continue;
                }
                
                
                break;

            case 15:
                //jl
                if (zf){
                    ip = (size_t)get_val(in.m1,in.v1);
                    continue;
                }
                
                
                break;
            
            case 16:
                //pri
                printf("%lld " , (int64_t)get_val(in.m1, in.v1));
                break;
            default:
                return -1;
                break;
        }

        ip++;

        /*
    "mov": 0,
    "inc": 1,
    "dec": 2,
    "neg": 3,  
    "shl": 4,  
    "shr": 5,  
    "add": 6,  
    "sub": 7,  
    "mul": 8,  
    "div": 9,
    "cmp": 10,
    "jmp": 11,
    "jne": 12,
    "je":  13,
    "jg":  14,
    "jl":  15,
    "pri": 16,
    */
        





    }
    clock_t stop = clock();
    double elapsed = (double)(stop - start) * 1000.0 / CLOCKS_PER_SEC;
    printf("\nTime elapsed in ms: %f", elapsed);
    
}